File: fem/tests/findnorm.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <string.h>

#define MAX_LENGTH 255

int main(int argc, char **argv)
{
   FILE *fp = fopen( argv[1], "r" );
   char line[MAX_LENGTH+1], *ptr;

   static double norm, f, target_nrm, target_eps = 1e-5, REALT, CPUT;
   int n, success;

   success = 0;
   while( fgets( line, MAX_LENGTH, fp ) )
   {
      if ( strstr( line, "END TEST CASE" ) ) {
         ptr = strstr( line, "NRM=" );         
         if ( ptr ) sscanf( ptr,"NRM=%lf", &target_nrm );
         ptr = strstr( line, "EPS=" );         
         if ( ptr ) sscanf( ptr,"EPS=%lf", &target_eps );         
         success = compare( norm, target_nrm, target_eps );
         if ( !success ) {
           n = strlen(line)-1;
           while( line[n]==10 || line[n]==13 ) line[n--] = '\0';
           fprintf( stderr, "[FAILED]: %s, Computed NRM=%g: ", line, norm );
           break;
         }
      }
      else if ( ptr=strstr( line, "(NRM,RELC)" ) ) {
        while( *ptr != '\0' && *ptr!='+' && *ptr != '-' && *ptr != '.' && !isdigit(*ptr) ) ptr++;
        n = sscanf( ptr, "%lf", &f );
        if ( n==1 && f != 0.0 ) norm = f;
      } else if (ptr = strstr( line, "Check NRM=") ) {
        while( *ptr != '\0' && *ptr!='+' && *ptr != '-' && *ptr != '.' && !isdigit(*ptr) ) ptr++;
        n = sscanf( ptr, "%lf", &f );
        if ( n==1 && f != 0.0 ) norm = f;
      } 
      else if ( ptr = strstr( line, "(CPU,REAL):" )  ) {
        sscanf(  ptr+11, "%lf %lf", &CPUT, &REALT); 
      }
   }

   if ( argc<=2 ) {
     fprintf( stdout, "%d\n", success );
   } else {
     fprintf( stdout, "%g\n", atof(argv[2])+CPUT );
   }
}



int compare( double norm1, double norm2, double eps )
{
   int n;

   if ( norm1 != -1 ) {
      if ( eps < 0 ) { 
         if ( norm2 < norm1 )
            return 1;
         else
            return 0;
      } else  if ( 2 * fabs(norm1-norm2) / (norm1+norm2) < eps )
         return 1;
      else
         return 0;
   } else {
      return 1;
   }
}

File: fem/tests/findnorm.f90
PROGRAM findnorm

 IMPLICIT NONE

 CHARACTER(LEN=512) :: line, cpustring, filename
 INTEGER :: Noargs, i,j,n, success
 REAL(8) :: target_nrm, target_eps = 1d-5, f, norm, CPUT, REALT, cumul_time

 NoArgs = IARGC() 
 CALL GetArg(1,filename)
 IF (NoArgs >= 2) THEN
   CALL GetArg(2,cpustring)
   READ(cpustring, *) cumul_time
 END IF

 OPEN( 1, File = filename, STATUS='old' )
 DO WHILE(.TRUE.)
   READ(1,'(a)', END=100, ERR=100 ) line
   i = INDEX( line, 'END TEST CASE' )
   IF (i>0 ) THEN
     i = INDEX( line, 'NRM=' )
     IF(i>0) READ(line(i+4:), * ) target_nrm

     i = INDEX( line, 'EPS=' )
     IF(i>0) READ(line(i+4:), * ) target_eps
   END IF

   i = INDEX(line, '(NRM,RELC)')
   IF(i>0) THEN
     DO j=i+10,LEN_TRIM(line)
       IF( line(j:j) == '(' ) THEN
         READ( line(j+1:), *, END=10,ERR=10 ) f
         IF(f/=0) norm = f
10       CONTINUE
       END IF
     END DO
   END IF

   i = INDEX(line, 'Check NRM')
   IF(i>0) THEN
     DO j=i+10,LEN_TRIM(line)
       IF( line(j:j) == '(' ) THEN
         READ( line(j+1:), *, END=20,ERR=20 ) f
         IF(f/=0) norm = f
20       CONTINUE
       END IF
     END DO
   END IF

   i = INDEX(line, '(CPU,REAL')
   IF (i>0) THEN
     READ(line(i+11:), *,END=30,ERR=30) REALT, CPUT
30   CONTINUE
   END IF
 END DO

100 CONTINUE

 Success = COMPARE( norm, target_nrm, target_eps );
 IF ( noargs < 2 ) THEN
   IF (Success<=0) THEN
     n = LEN_TRIM(line)
     j = n
     DO i=n,1,-1
       IF(ICHAR(line(i:i))==10.OR.ICHAR(line(i:i))==13) j=j-1
     END DO
     WRITE(6,'(a,g0.6,a,g0.6)' ) '[FAILED] Computed NRM=', norm,' Target NRM=', target_nrm
    ELSE
     WRITE(6,'(I1)') Success
    END IF
 ELSE
   WRITE(6, '(g10.4)') cumul_time + CPUT
 END IF


  FLUSH(6)

CONTAINS

FUNCTION Compare( norm1, norm2, eps ) RESULT(success)
  REAL(8) :: norm1, norm2, eps
  INTEGER :: success
 
  success = 0

  if ( norm1 /= -1 )then
    if ( eps < 0 ) then
       if ( norm2 < norm1 ) then
          success = 1
          return;
       else
          return;
       end if
    else  if ( 2 * abs(norm1-norm2) / (norm1+norm2) < eps ) then
       success = 1
       return
    else
       return
    end if
  else 
    success = 1
    return
  end if
END FUNCTION Compare

END PROGRAM findnorm
